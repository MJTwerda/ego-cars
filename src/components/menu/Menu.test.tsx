import { act, fireEvent, render, RenderResult, waitFor } from "@testing-library/react";
import { useRouter } from "next/navigation";
import CommonMenu, { CommonMenuProps } from "./Menu";
import CloseIcon from '@mui/icons-material/Close';

const mockHandleClick = jest.fn();

const mockProps: CommonMenuProps = {
  menuLabel: {
    label: 'Menu',
  },
  menuIconClosed: <CloseIcon />,
  menuItems: [
    {
      label: 'Item 1',
      path: '/',
    },
    {
      label: 'Item 2',
      bottomDivider: true,
      handleClick: mockHandleClick
    },
    {
      label: 'Item 3',
      path: 'item-3'
    }
  ],
  justifyItems: 'start'
};

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

const mockPush = jest.fn();
(useRouter as jest.Mock).mockReturnValue({ push: mockPush });

describe('Menu', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render', async () => {
    let component: RenderResult;
    await act(async () => {
      component = render(<CommonMenu {...mockProps} />);
    });

    await waitFor(() => {
      expect(component).toMatchSnapshot();
    });
  });

  it('should render with all props', async () => {
    const newProps: CommonMenuProps = {
      ...mockProps,
      menuLabel: {
        ...mockProps.menuLabel,
        isStrong: true
      },
      menuIconOpened: <></>,
      showCloseButton: true,
      justifyItems: 'end'
    }
    let component: RenderResult;
    await act(async () => {
      component = render(<CommonMenu {...newProps} />);
    });

    await waitFor(() => {
      expect(component).toMatchSnapshot();
      const iconButton = component.getByTestId('common_menu_icon_btn');
      fireEvent.click(iconButton);

      expect(component.getByRole('menu')).toBeInTheDocument();
      fireEvent.click(iconButton);
    });
  });

  it('should close clicking on Menu Icon button', async () => {
    const newProps: CommonMenuProps = {
      ...mockProps,
      showCloseButton: true,
    }
    let component: RenderResult;

    await act(async () => {
      component = render(<CommonMenu {...newProps} />);
    });

    await waitFor(() => {
      fireEvent.click(component.getByTestId('common_menu_icon_btn'));
      expect(component.getByTestId('common_menu_close_btn')).toBeInTheDocument();
    });
  });

  it('should show close button option if showCloseButton is true', async () => {
    const newProps: CommonMenuProps = {
      ...mockProps,
      showCloseButton: true,
    }
    let component: RenderResult;

    await act(async () => {
      component = render(<CommonMenu {...newProps} />);
    });

    await waitFor(() => {
      fireEvent.click(component.getByTestId('common_menu_icon_btn'));

      const closeBtnOption = component.getByTestId('common_menu_close_btn');
      expect(closeBtnOption).toBeInTheDocument();

      fireEvent.click(closeBtnOption);
    });

  });

  it('should navigate clicking on Link option', async () => {
    const newProps: CommonMenuProps = {
      ...mockProps,
      showCloseButton: true,
    }
    let component: RenderResult;

    await act(async () => {
      component = render(<CommonMenu {...newProps} />);
    });

    await waitFor(() => {
      fireEvent.click(component.getByTestId('common_menu_icon_btn'));

      const navigateBtn = component.getByTestId('common_menu_item_Item 3');
      expect(navigateBtn).toBeInTheDocument();

      const menuItemLink = component.getByTestId('common_menu_item_link_/');
      expect(menuItemLink).toHaveAttribute('href', '/');

      fireEvent.click(navigateBtn);
    });

  });

  it('should execute option function clicking on option', async () => {
    const newProps: CommonMenuProps = {
      ...mockProps,
      showCloseButton: true,
    }
    let component: RenderResult;

    await act(async () => {
      component = render(<CommonMenu {...newProps} />);
    });

    await waitFor(() => {
      fireEvent.click(component.getByTestId('common_menu_icon_btn'));

      const option2 = component.getByTestId('common_menu_item_Item 2');
      expect(option2).toBeInTheDocument();

      fireEvent.click(option2);

      expect(mockHandleClick).toHaveBeenCalledTimes(1);
    });

  });
})