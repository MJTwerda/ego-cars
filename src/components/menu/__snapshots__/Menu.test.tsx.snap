// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="flex justify-center"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 flex self-center w-3/4 css-rizt0-MuiTypography-root"
          >
            Menu
          </p>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
            data-testid="common_menu_icon_btn"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="flex justify-center"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 flex self-center w-3/4 css-rizt0-MuiTypography-root"
        >
          Menu
        </p>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
          data-testid="common_menu_icon_btn"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Menu should render with all props 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="flex justify-center"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 flex self-center w-3/4 css-rizt0-MuiTypography-root"
          >
            <strong>
              Menu
            </strong>
          </p>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
            data-testid="common_menu_icon_btn"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="flex justify-center"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 flex self-center w-3/4 css-rizt0-MuiTypography-root"
        >
          <strong>
            Menu
          </strong>
        </p>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
          data-testid="common_menu_icon_btn"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
